Authorization checks

marlonbaeten opened the pull request:
- Adds authorization to backend endpoints via `FromRequestParts` of a `User`. Newtypes around `User`, e.g. Typist(pub User), provide separation of roles. 
- API requests in backend tests now provide a session cookie where needed
- Updated OpenAPI spec with 401 response where applicable
- API requests in e2e tests now provide session cookies with the correct roles
- You can login as different roles via the dev UI
- User and role are now shown in the navigation bar

To quickly login as one of the roles, use the buttons on the `/dev` page:

![image](https://github.com/user-attachments/assets/b6acf30d-6f68-46a8-85dc-56aa032e5d2d)

Applicable links will appear on this page once logged in.

To test:

Use the application as before, but login as one of the roles. Try to find unprotected API endpoints! There are still front-end routes that do render when not logged in as the applicable role, but their API call's will result in an error.

Note that this is a large PR, but a lot of the issues listed below are closely related or have co-dependencies.

Fixes #999
Fixes #1000
Fixes #1001
Fixes #1003
Fixes #1004
Fixes #1005

praseodym commented:
The links to `/users` and `/workstations` should be hidden from the navbar (links and menu) for the coordinator role.

oliver3 commented on frontend/app/module/election/page/ElectionHomePage.tsx:
```diff
@@ -1,14 +1,19 @@
+  const user = useUser();
   const { election } = useElection();
 
+  if (user?.role === "typist") {
```
It seems that `useUser` is only used to check for the user role? Maybe we can repurpose it to something like 
```ts
const {isTypist, isAdministrator, isCoordinator } = useUserRole()
```

marlonbaeten replied:
Fixed in 29d2235

marlonbaeten commented on frontend/e2e-tests/state/admin.json:
```diff
@@ -0,0 +1,15 @@
+{
```
Fixed

marlonbaeten commented on backend/openapi.json:
```diff
@@ -2164,9 +2365,17 @@
         "required": [
           "user_id",
-          "username"
+          "fullname",
```
Fixed in 29d2235

marlonbaeten commented on backend/src/authentication/api.rs:
```diff
@@ -28,14 +28,21 @@ pub struct Credentials {
     fn from(user: &User) -> Self {
         Self {
             user_id: user.id(),
+            fullname: user
```
Fixed in 29d2235

oliver3 commented on backend/src/authentication/api.rs:
```diff
@@ -28,7 +28,8 @@ pub struct Credentials {
 pub struct LoginResponse {
     pub user_id: u32,
-    pub fullname: String,
+    #[schema(value_type = String, nullable = true)]
```
```suggestion
    #[schema(value_type = String, nullable = false)]
```

marlonbaeten replied:
Fixed in 293485a

oliver3 commented on backend/src/authentication/user.rs:
```diff
@@ -26,7 +26,7 @@ pub struct User {
     username: String,
     #[serde(skip_serializing_if = "Option::is_none")]
-    #[schema(nullable = false)]
+    #[schema(value_type = String, nullable = true)]
```
```suggestion
    #[schema(value_type = String, nullable = false)]
```

praseodym reviewed (changes requested):
The frontend doesn't handle authorization errors well. I'd expect to be redirected to the login page instead of seeing an empty page or error page. For example:
- `/elections` => error page, 401
- `/account/change-password` => empty page

Furthermore, the 'Afmelden' navbar link logs out immediately which doesn't work well on the data entry form. When I click 'Afmelden', the 'Wat wil je doen met je invoer?' modal is shown but the session is already terminated. Clicking 'Invoer bewaren' or 'Niet bewaren' in the modal no longer works at that point.

marlonbaeten commented:
> The frontend doesn't handle authorization errors well. I'd expect to be redirected to the login page instead of seeing an empty page or error page. For example:
> 
> * `/elections` => error page, 401
> * `/account/change-password` => empty page
> 
> Furthermore, the 'Afmelden' navbar link logs out immediately which doesn't work well on the data entry form. When I click 'Afmelden', the 'Wat wil je doen met je invoer?' modal is shown but the session is already terminated. Clicking 'Invoer bewaren' or 'Niet bewaren' in the modal no longer works at that point.

Thanks for your review!

I have added a redirect to the login page if the error boundary catches a 401 error and added explicit checks+redirects for pages where there is no API request performed. This should solve your first issue.

The second issue is solved by having an explicit logout route and component. The component will logout the user only after navigating.

See commit 7f43bb7da9ab0f5b158f82cf35f8485d62394803

praseodym reviewed (approved):
lgtm, test coverage could be a bit higher but is ok for now
